sumpy/e2e.py

```diff
@@ -251,6 +330,113 @@ def __call__(self, queue, **kwargs):
+                for i, coeff_i in enumerate(
+                    self.tgt_expansion.m2l_global_precompute_exprs(
+                        self.src_expansion, src_rscale,
+                        dvec, tgt_rscale, sac))]
+
+        sac.run_global_cse()
+
+        from sumpy.codegen import to_loopy_insns
+        return to_loopy_insns(
+                sac.assignments.items(),
+                vector_names=set(["d"]),
+                pymbolic_expr_maps=[self.tgt_expansion.get_code_transformer()],
+                retain_names=tgt_coeff_names,
+                complex_dtype=np.complex128  # FIXME
+                )
+
+    def get_kernel(self):
+        nprecomputed_exprs = \
+            self.tgt_expansion.m2l_global_precompute_nexpr(self.src_expansion)
+        from sumpy.tools import gather_loopy_arguments
+        loopy_knl = lp.make_kernel(
+                [
+                    "{[itr_class]: 0<=itr_class<ntranslation_classes}",
+                    "{[idim]: 0<=idim<dim}",
+                    ],
+                ["""
+                for itr_class
+                    <> d[idim] = m2l_translation_vectors[idim, \
+                            itr_class + translation_classes_level_start] {dup=idim}
+
+                    """] + self.get_translation_loopy_insns() + ["""
+                    m2l_precomputed_exprs[itr_class, {idx}] = precomputed_expr{idx}
+                    """.format(idx=i) for i in range(nprecomputed_exprs)] + ["""
+                end
+                """],
+                [
+                    lp.ValueArg("src_rscale", None),
+                    lp.GlobalArg("m2l_precomputed_exprs", None,
+                        shape=("ntranslation_classes", nprecomputed_exprs),
+                        offset=lp.auto),
+                    lp.GlobalArg("m2l_translation_vectors", None,
+                        shape=("dim", "ntranslation_vectors")),
+                    lp.ValueArg("ntranslation_classes", np.int32),
+                    lp.ValueArg("ntranslation_vectors", np.int32),
+                    lp.ValueArg("translation_classes_level_start", np.int32),
+                    "..."
+                ] + gather_loopy_arguments([self.src_expansion, self.tgt_expansion]),
+                name=self.name,
+                assumptions="ntranslation_classes>=1",
+                default_offset=lp.auto,
+                fixed_parameters=dict(dim=self.dim),
+                lang_version=MOST_RECENT_LANGUAGE_VERSION
+                )
+
+        for knl in [self.src_expansion.kernel, self.tgt_expansion.kernel]:
+            loopy_knl = knl.prepare_loopy_kernel(loopy_knl)
+
+        loopy_knl = lp.tag_inames(loopy_knl, "idim*:unr")
+        loopy_knl = lp.tag_inames(loopy_knl, dict(idim="unr"))
+
+        return loopy_knl
+
+    def get_optimized_kernel(self):
+        # FIXME
+        knl = self.get_kernel()
+        knl = lp.split_iname(knl, "itr_class", 16, outer_tag="g.0")
```

Comment:
How many `itr_class` entries do you expect there to be? Is this an even minimally effective parallelization strategy? Wouldn't `tag_inames(knl, {"itr_class": "g.0"})` be better?